LinkedList/Exercise/SLList.h

```diff
@@ -0,0 +1,328 @@
+
+		if (ListSize == 0)
+		{
+			Tail = Head;
+		}
+
+		ListSize++;
+	}
+
+	//-------------------------------------------------------------------------------------
+
+	void AddTail(const Type& InElement)
+	{
+		if (ListSize == 0)
+		{
+			AddHead(InElement);
+		}
+		else
+		{
+			FNode* NewTail = new FNode();
+			NewTail->Element = InElement;
+
+			Tail->Next = NewTail;
+			Tail = NewTail;
+
+			ListSize++;
+		}
+
+	}
+
+	//-------------------------------------------------------------------------------------
+
+	void Insert(const Type& InElement, const int Index)
+	{
+		if (Index == 0)
+		{
+			AddHead(InElement);
+		}
+		else if (Index > 0 && Index < ListSize)
+		{
+			FNode* Current = GetNode(Index);
+
+			FNode* NewNode = new FNode();
+			NewNode->Element = InElement;
+
+			GetNode(Index - 1)->Next = NewNode;
+			NewNode->Next = Current;
+			ListSize++;
+		}
+	}
+
+	//-------------------------------------------------------------------------------------
+
+	void Remove(const int Index)
+	{
+		if (Index == 0)
+		{
+			FNode* NextUp = Head->Next;
+			delete Head;
+			Head = NextUp;
+		}
+		else if (Index > 0 && Index < ListSize)
+		{
+			FNode* Previous = GetNode(Index - 1);
+			FNode* Pending = Previous->Next;
+			Previous->Next = Pending->Next;
+			delete Pending;
+			ListSize--;
+		}
+	}
+
+	//-------------------------------------------------------------------------------------
+
+	void Clear()
+	{
+		FNode* NextUp;
+
+		for (int i = 0; i < ListSize; i++)
+		{
+			NextUp = Head->Next;
+			delete Head;
+			Head = NextUp;
+		}
+		ListSize = 0;
+	}
+
+	//-------------------------------------------------------------------------------------
+
+	template<typename Pred>
+	void ForEach(const Pred& Predicate)
+	{
+		for (int i = 0; i < ListSize; i++)
+		{
+			Predicate(GetNode(i));
+		}
+	}
+
+	//-------------------------------------------------------------------------------------
+
+	template<typename Pred>
+	Type* FindByPredicate(const Pred& Predicate) const
+	{
+		FNode* Current = Head;
+		while (Current)
+		{
+			if (Predicate(Current->Element))
+			{
+				return &Current->Element;
+			}
+
+			Current = Current->Next;
+		}
+
+		//if doesn't find the element
+		return nullptr;
+	}
+
+	//-------------------------------------------------------------------------------------
+
+	template<typename Pred>
+	TSLList<Type> FilterByPredicate(const Pred& Predicate) const
+	{
+		TSLList<Type> NewList;
+
+		FNode* Current = Head;
+		for (int i = 0; i < ListSize; i++)
+		{
+			if (Predicate(Current->Element))
+			{
+				NewList.AddTail(Current->Element);
+			}
+			Current = Current->Next;
+		}
+		return NewList;
+	}
+
+	//-------------------------------------------------------------------------------------
+
+	template<typename Pred>
+	void RemoveAllByPredicate(const Pred& Predicate) const
+	{
+		FNode* Current = Head;
+		for (int i = 0; i < ListSize; i++)
+		{
+			if (Predicate(Current->Element))
+			{
+				Remove(i);
+			}
+			Current = Current->Next;
```

Comment:
This might not work as desired. If you remove the i-th element, say i=3, the element that used to be at i=4 will now be where the one you removed used to be, so calling remove(4) will not to what you're expecting it to do. Also, it might not to absolutely nothing since ListSize is decreasing on each iteration, so you might be calling remove(4) on a list of 4 elements.

Reply:
fixed 👌